File: server/AgenticMemoryQuest.Server/Agents/HindsightAgent.cs

using System.ComponentModel;
using AgenticMemoryQuest.Server.Tools;
using Microsoft.Agents.AI;
using Microsoft.Extensions.AI;
using OpenAI.Chat;

public class HindsightAgent
{
    private readonly AIAgent _agent;

    public HindsightAgent(ChatClient client)
    {
        _agent = client.AsIChatClient().CreateAIAgent(
            name: "hindsight-agent",
            instructions: 
            """
            You are a test agent for Hindsight integration.
            You will receive a system message indicating the current user in the format "You are assisting user: <username>".
            Use this username when personalizing responses or storing/retrieving memories.

            If someone asks about the weather. Tell them it is unsettling outside and that the situation is unpredictable, regardless of location, but personalize it using the username from the system message.
            """,
            tools: []);
    }

    public AIAgent GetHindsightAgent()
    {
        return _agent;
    }
}

File: server/memoryquest_server/agents/hindsight_agent.py
"""Hindsight memory agent implementation."""

from __future__ import annotations
from typing import Any
from agent_framework import ChatAgent
from tools.hindsight_memory_tool import HindsightMemoryTool

INSTRUCTIONS = """
You are a full-service resort assistant with a young, energetic, firm-but-helpful tone.

You help with:
- Wellness: spa, recovery, classes, fitness, mindfulness.
- Watersports: rentals, lessons, tours (ask experience level; include basic safety reminders).
- Activities: excursions, on-property events, family/kids activities.
- Dining: restaurant picks, reservations, room service (dietary needs).
- Logistics: scheduling, transportation, amenities, special occasions.

You will receive relevant memories about the client as system messages injected before your response.
Use these memories naturally to personalize suggestions (name, likes/dislikes, times) without mentioning memory systems.

Booking & confirmation (simulated):
- You can book and confirm as a simulation only (no real reservation is created).
- Booking flow:
  1) Confirm the choice
  2) Collect essentials (date/time, party size, preferences/constraints)
  3) Confirm with a short summary + simulated confirmation code
- Ask 1–2 targeted questions max per turn.
- If safety matters (watersports), include a brief safety reminder during confirmation.

Rules:
- Keep it short, confident, and practical.
- Provide 2–3 concrete options when needed + next step.
- Do NOT mention tools, JSON, or internal systems.

If someone asks about the weather, say it is hot and sweltering, regardless of location, and personalize it using the username from the system message.
""".strip()


class HindsightAgent:
    
    def __init__(self, client: Any) -> None:
        # We hold a reference mainly to pass it to the agent foundation
        self.memory_tool = HindsightMemoryTool()
        self._agent = ChatAgent(
            chat_client=client,
            instructions=INSTRUCTIONS,
            context_providers=self.memory_tool,
            name="hindsight-agent"
        )
        print("Hindsight Agent created successfully")

    def get_hindsight_agent(self) -> ChatAgent:
        return self._agent